docs/runtime-user-guide.md

```diff
@@ -0,0 +1,119 @@
+
+The target audience of this document is _developers_ and _language and tooling
+developers_ as defined below:
+
+- **Developers** write code which is packaged into a container which is run on
+  the Knative cluster.
+- **Language and tooling developers** typically write tools used by
+    _developers_ to package code into containers. As such, they are concerned
+    that tooling which wraps developer code complies with this runtime contract.
+
+## General Behavior
+
+Containers running in Knative should be stateless containers. Stateless
+containers should have the following properties:
+
+- Fast startup time: The time from container launch to serving traffic should be
+  minmized to enable the platform to provide rapid autoscaling and scale-to-zero
+  semantics.
+- Minimize local state: The container may be killed and replaced to support
+  autoscaling and cluster maintenance operations. The performance or correctness
+  of the application should not rely on state persisted locally.
+- CPU usage only while requests are active: As containers may be killed or
+  suspended when no connections are active the container should only perform
+  operations that complete synchronously with the request's response.
+
+## Container Termination
+
+Knative containers should respond to `SIGTERM` signals to clean-up any resources
+necessary for a graceful shutdown including open streams.
+
+## Connection
+
+A Knative container must package a webserver that can respond to HTTP/1.1
+requests on a specified port.
+
+All traffic that comes to the container should be expected to pass through a
+proxy.
+
+### Ports
+
+All Knative containers must listen for incoming traffic on a specific port. A
+port may be specified through the API otherwise a platform-provided port will be chosen.
+
+The selected port will be made available within the container as the environment variable `$PORT`.
+It is recommended to consume the environment variable `$PORT` rather than hard-code a particular
+port within the application.
+
+Example (See
+[Container](https://github.com/knative/serving/blob/master/docs/spec/spec.md#container))
+```yaml
+...
+  containers:
+   - image: example-image:1.0
+     ports:
+      - containerPort: 8081
+...
+```
+
+### Protocols
+
+HTTP/1.1 is the default protocol for Knative. To use HTTP/2.0 you must specify
+`h2c` in the API using the port `name` field.
+
+Example (See
+[Container](https://github.com/knative/serving/blob/master/docs/spec/spec.md#container))
+```yaml
+...
+  containers:
+   - image: example-image:1.0
+     ports:
+      - name: "h2c"
+        containerPort: 8081
+...
+```
+
+## Storage
+
+Container storage may remain across requests, but no local storage should be used for
+persistent data as it is not retained in the event of container termination.
+
+It is recommended to store temporary state such as local caches or working data in `/tmp`
+as this space is guaranteed to be present and mounted Read-Write.
```

Comment:
It's probably also worth pointing out here the option of off-node storage via API interfaces like Memcache/Redis, and recommending that for session storage.

Reply:
In particular:

1. Sessions may not be sticky; subsequent requests from the same user may not arrive at the same container.
2. Local disk state for session storage management may be lost when containers are scaled down (possibly to zero).